proposals/trestle-sdk-cli-decoupling.md

```diff
@@ -0,0 +1,160 @@
+
+Below is high-level description of what would be moved:
+
+**OSCAL Classes**: The classes that represet the OSCAL data model.  
+**Core Functionality**: Contains the fundamental APIs for interacting with OSCAL objects.  
+**Abstractions**: Abstractions for common operations like resolution or core OSCAL validation.  
+  
+</details>
+
+## Design Details
+
+Below is diagram with existing `trestle` logic that is proposed to be part of the SDK. Links to where this type of logic exists currently are included:
+
+```mermaid
+classDiagram
+    OSCALBaseModel <|-- Catalog
+    OSCALBaseModel <|-- Profile
+    OSCALBaseModel <|-- ComponentDefinitions
+    OSCALBaseModel <|-- SSP
+    OSCALBaseModel <|-- AssessmentPlan
+    OSCALBaseModel <|-- AssessmentResult
+    OSCALBaseModel <|-- POAM
+    class OSCALBaseModel{
+      +Config
+      +write()
+      +read()
+    }
+    class Catalog{
+      +CatalogAPI
+    }
+    class Validator{
+      +isValid(OSCALBaseModel)
+    }
+    class Resolver{
+      +Fetcher
+      +import()
+      +merge()
+      +modify()
+      +prune()
+    }
+```
+
+### Relevant Links
+
+ - OSCAL Classes: https://github.com/oscal-compass/compliance-trestle/tree/develop/trestle/oscal
+  - Core Logic: https://github.com/oscal-compass/compliance-trestle/blob/develop/trestle/core/models/
+  - Abstractions: 
+    - https://github.com/oscal-compass/compliance-trestle/tree/develop/trestle/core/resolver
+    - https://github.com/oscal-compass/compliance-trestle/blob/dfe892936e5960ad64f6f387dbe5918314049e89/trestle/core/validator.py
+
+## Impacts / Key Questions
+
+Some of my main questions around this proposal are the following:
+
+- What will the contribution process look like for logic in `trestle`?
+- Will the SDK be maintained by the `compliance-trestle` maintainer or by a new set of maintainers?
+
+### Pros
+
+- **Improved maintainability**: Easier to manage and update individual components.
+- **Enhanced reusability**: SDK can be used in more applications and simplify contributions.  
+- **Reduced risk of breaking changes**: Fewer unintended consequences around codebase changes.
+
+### Cons
+
+- **Increased complexity**: Managing two separate repositories mean more project infrastructure to manage.
+- **Initial effort**: Moving the SDK to a new repository requires upfront work.
+
+## Risks and Mitigations
+
+- **User Impact**: Breaking changes and movement will affect current SDK users and this will require community notifications and discussion as well as migration documentation.
+
+### Security Considerations
+
+- **Increased Attack Surface**: Managing more repositories and  distinct codebases can potentially increase the attack surface.
+
+## Future Milestones (Optional)
+
+- Add a Go SDK to better support the C2P Go version and expand potential for integrations in the cloud native space.
```

Comment:
What do you think to take a phasing approach as initially creating Go binding and then SDK written in pure Go if requested?
The Go binding consists of OSCAL models (Go struct of Catalog/Profile/CDef/SSP/SAP/SAR/PoAM), and interfaces of OSCAL Base Model and validators. The interfaces are bound to Python SDK as invoking the same functions of Python SDK.

Pros: Development cost
Cons: Install Trestle-SDK and Python

Reply:
Thanks for bringing this up @yana1205. This is something I have considered, but it has not been discussed. I found it be more straight forward to call Go from Python instead of the other way around. On top of the cons you listed for the binding solution, complexity may add maintenance costs as well IMO.

As I understand it, we have an initial use case for Go from the C2P perspective (through the use case you mentioned below and through some of [re-implementation](https://github.com/oscal-compass/compliance-to-policy/tree/main/go/pkg/types/oscal) in the C2P `/go` directory). I think we could explore the trade-offs in more details in a follow-on proposal.

To confirm, does the outcome of binding vs porting affect your thoughts on this proposal which was just SDK separation?